driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -16,168 +16,125 @@
-            if (connection != null)
-                connection.release();
-            logError(host.getSocketAddress(), e);
-            return false;
-        } catch (BusyConnectionException e) {
-            // The pool shouldn't have give us a busy connection unless we've maxed up the pool, so move on to the next host.
-            connection.release();
-            logError(host.getSocketAddress(), e);
-            return false;
-        } catch (TimeoutException e) {
-            // We timeout, log it but move to the next node.
-            logError(host.getSocketAddress(), new DriverException("Timeout while trying to acquire available connection (you may want to increase the driver number of per-host connections)"));
-            return false;
-        } catch (RuntimeException e) {
-            if (connection != null)
-                connection.release();
-            logger.error("Unexpected error while querying " + host.getAddress(), e);
-            logError(host.getSocketAddress(), e);
-            return false;
-        }
+        SpeculativeExecution execution = new SpeculativeExecution(++executionCount);
+        runningExecutions.add(execution);
+        execution.sendRequest();
     }
 
-    private void write(Connection connection, Connection.ResponseCallback responseCallback) throws ConnectionException, BusyConnectionException {
-        // Make sure cancel() does not see a stale connectionHandler if it sees the new query state
-        // before connection.write has completed
-        connectionHandler = null;
-
-        // Ensure query state is "in progress" (can be already if connection.write failed on a previous node and we're retrying)
-        while (true) {
-            QueryState previous = queryStateRef.get();
-            if (previous.isCancelled()) {
-                if (connection instanceof PooledConnection)
-                    ((PooledConnection)connection).release();
-                return;
-            }
-            if (previous.inProgress || queryStateRef.compareAndSet(previous, previous.startNext()))
-                break;
-        }
-
-        connectionHandler = connection.write(responseCallback, false);
-        // Only start the timeout when we're sure connectionHandler is set. This avoids an edge case where onTimeout() was triggered
-        // *before* the call to connection.write had returned.
-        connectionHandler.startTimeout();
-
-        // Note that we could have already received the response here (so onSet() / onException() would have been called). This is
-        // why we only test for CANCELLED_WHILE_IN_PROGRESS below.
+    private void scheduleExecution(long delayMillis) {
+        if (isDone.get() || delayMillis <= 0)
+            return;
+        logger.trace("[{}] Schedule next speculative execution in {} ms", id, delayMillis);
+        scheduledExecutions.add(scheduler.newTimeout(newExecutionTask, delayMillis, TimeUnit.MILLISECONDS));
```

Comment:
That could happen with an extremely low speculative delay, which is unlikely. But anyway, leaving a completed `Timeout` in `scheduledExecutions` is not really a problem, the worst that can happen is that we cancel it when it's done, which is a no-op.